ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -31,215 +81,264 @@ where
+        mut uv: J,
+    ) -> Self
+    where
+        F: Default,
+        ZeroKnowledgeProof<F, R>: ProofArray,
+        J: Iterator<Item = B>,
+        B: Borrow<UVPolynomial<F, R>>,
+    {
+        let lagrange_table_r = Self::retrieve_lagrange_table(proof_left, proof_right);
+
+        let mut output = Vec::<UVPolynomial<F, R>>::new();
+
+        // iter over chunks of size M
+        // and interpolate at x coordinate r
+        while let Some(polynomial) = uv.next() {
+            let mut u = GenericArray::<F, R>::default();
+            let mut v = GenericArray::<F, R>::default();
+            u[0] = lagrange_table_r.eval(&polynomial.borrow().0)[0];
+            v[0] = lagrange_table_r.eval(&polynomial.borrow().1)[0];
+            for i in 1..R::USIZE {
+                if let Some(polynomial) = uv.next() {
+                    u[i] = lagrange_table_r.eval(&polynomial.borrow().0)[0];
+                    v[i] = lagrange_table_r.eval(&polynomial.borrow().1)[0];
+                } else {
+                    u[i] = F::ZERO;
+                    v[i] = F::ZERO;
+                }
             }
-            let p_extrapolated = lagrange_table.eval(&p);
-            let q_extrapolated = lagrange_table.eval(&q);
+            output.push((u, v));
+        }
 
-            for (i, (x, y)) in
-                zip(p_extrapolated.into_iter(), q_extrapolated.into_iter()).enumerate()
-            {
-                proof[λ::USIZE + i] += x * y;
-            }
+        Self { uv: output }
+    }
+
+    /// This function computes a `LagrangeTable` for a `Fiat-Shamir` challenge
+    /// which is computed via hashing `proof_left` and `proof_right`
+    fn retrieve_lagrange_table(
+        proof_left: &ZeroKnowledgeProof<F, R>,
+        proof_right: &ZeroKnowledgeProof<F, R>,
+    ) -> LagrangeTable<F, R, U1>
+    where
+        ZeroKnowledgeProof<F, R>: ProofArray,
+    {
+        let r: F = hash_to_field(
+            &compute_hash(proof_left.g.iter()),
+            &compute_hash(proof_right.g.iter()),
+            R::U128,
+        );
+
+        let denominator = CanonicalLagrangeDenominator::<F, R>::new();
+        LagrangeTable::<F, R, U1>::new(&denominator, &r)
+    }
+}
+
+#[cfg(all(test, unit_test))]
+mod test {
+    use std::{iter::zip, ops::Add};
+
+    use generic_array::{sequence::GenericSequence, ArrayLength, GenericArray};
+    use typenum::{Sum, U1, U2, U3, U4, U7};
+
+    use super::UVStore;
+    use crate::{
+        ff::{Fp31, PrimeField, U128Conversions},
+        protocol::ipa_prf::malicious_security::{
+            lagrange::{CanonicalLagrangeDenominator, LagrangeTable},
+            prover::{TwoNPlusOne, UVPolynomial, ZeroKnowledgeProof},
+        },
+        secret_sharing::SharedValue,
+    };
+
+    impl Default for Fp31 {
+        fn default() -> Self {
+            Self::ZERO
         }
-        ZeroKnowledgeProof::new(proof)
     }
 
-    pub fn compute_final_proof<λ, I, J>(
-        u: I,
-        v: J,
+    // todo: deprecate
```

Comment:
Why does this need to be deprecated?

Reply:
final proof is no longer needed. We don't need two different functions for the final and intermediate proofs. I left it there since your tests still use them.